clippy_lints/src/macro_use.rs

```diff
@@ -0,0 +1,55 @@
+use crate::utils::span_lint_and_help;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use syntax::ast;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for `#[macro_use] use...`.
+    ///
+    /// **Why is this bad?** Since the Rust 2018 edition you can import
+    /// macro's directly, this is considered idiomatic.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// #[macro_use]
+    /// use lazy_static;
+    /// ```
+    pub MACRO_USE_IMPORT,
+    pedantic,
+    "#[macro_use] is no longer needed"
+}
+
+declare_lint_pass!(MacroUseImport => [MACRO_USE_IMPORT]);
+
+impl EarlyLintPass for MacroUseImport {
+    fn check_item(&mut self, ecx: &EarlyContext<'_>, item: &ast::Item) {
+        if let ast::ItemKind::Use(use_tree) = &item.kind {
+            let macro_use_attr = item
+                .attrs
+                .iter()
+                .find(|attr| attr.ident().map(|s| s.to_string()) == Some("macro_use".to_string()));
+            if let Some(mac_attr) = macro_use_attr {
+                let msg = "`macro_use` attribute's are no longer needed in the Rust 2018 edition";
+                let help = format!(
+                    "remove the attribute and import the macro directly `use {}::macro_name`",
+                    use_tree
+                        .clone()
+                        .into_inner()
+                        .prefix
+                        .segments
+                        .iter()
+                        .enumerate()
+                        .map(|(i, s)| if i == 0 {
+                            s.ident.to_string()
+                        } else {
+                            format!("::{}", s.ident)
+                        })
+                        .collect::<String>(),
+                );
+                span_lint_and_help(ecx, MACRO_USE_IMPORT, mac_attr.span, msg, &help);
```

Comment:
The current implementation doesn't consider edition, it's also triggered on 2015 edition.
Add this condition:
```rust
if cx.sess.opts.edition == Edition::Edition2018 {}
```
Also you should add `// edition:2018` flag to the test to use 2018 edition.

Reply:
Ah yea good call thanks!